src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/BlockBalancedTreeWriter.java

```diff
@@ -0,0 +1,766 @@
+        if (reader.needsSorting())
+            MutablePointsReaderUtils.sort(Math.toIntExact(maxDoc), bytesPerValue, reader, 0, Math.toIntExact(reader.size()));
+
+        OneDimensionBKDWriter oneDimWriter = new OneDimensionBKDWriter(out, callback);
+
+        reader.intersect((docID, packedValue) -> oneDimWriter.add(packedValue, docID));
+
+        long filePointer = oneDimWriter.finish();
+
+        SAICodecUtils.writeFooter(out);
+        return filePointer;
+    }
+
+    interface Callback
+    {
+        void writeLeafDocs(RowIDAndIndex[] leafDocs, int offset, int count);
+    }
+
+    public static class RowIDAndIndex
+    {
+        public int valueOrderIndex;
+        public long rowID;
+
+        @Override
+        public String toString()
+        {
+            return MoreObjects.toStringHelper(this)
+                              .add("valueOrderIndex", valueOrderIndex)
+                              .add("rowID", rowID)
+                              .toString();
+        }
+    }
+
+    private class OneDimensionBKDWriter
+    {
+        private final IndexOutput out;
+        private final List<Long> leafBlockFilePointer = new ArrayList<>();
+        private final List<byte[]> leafBlockStartValues = new ArrayList<>();
+        private final byte[] leafValues = new byte[maxPointsInLeafNode * bytesPerValue];
+        private final long[] leafDocs = new long[maxPointsInLeafNode];
+        private final RowIDAndIndex[] rowIDAndIndexes = new RowIDAndIndex[maxPointsInLeafNode];
+        private final int[] orderIndex = new int[maxPointsInLeafNode];
+        private final Callback callback;
+        private final GrowableByteArrayDataOutput scratchOut = new GrowableByteArrayDataOutput(32 * 1024);
+        private final GrowableByteArrayDataOutput scratchOut2 = new GrowableByteArrayDataOutput(2 * 1024);
+        private final byte[] lastPackedValue = new byte[bytesPerValue];
+
+        private long valueCount;
+        private int leafCount;
+        private long lastDocID;
+
+        OneDimensionBKDWriter(IndexOutput out, Callback callback)
+        {
+            this.out = out;
+            this.callback = callback;
+
+            for (int x = 0; x < rowIDAndIndexes.length; x++)
+            {
+                rowIDAndIndexes[x] = new RowIDAndIndex();
+            }
+        }
+
+        void add(byte[] packedValue, long docID) throws IOException
+        {
+            if (DEBUG)
+                valueInOrder(valueCount + leafCount, lastPackedValue, packedValue, 0, docID, lastDocID);
+
+            System.arraycopy(packedValue, 0, leafValues, leafCount * bytesPerValue, bytesPerValue);
+            leafDocs[leafCount] = docID;
+            leafCount++;
+
+            if (leafCount == maxPointsInLeafNode)
+            {
+                // We write a block once we hit exactly the max count ... this is different from
+                // when we write N > 1 dimensional points where we write between max/2 and max per leaf block
```

Comment:
nit: I guess we don't even have to mention the `N > 1` case here...because it's just not a thing?